### ISSUE TITLE
GRPOTrainer: liger_kernel loss implementation ignores PEFT/LoRA adapters on lm_head

### ISSUE THREAD

**Issue by junho328** [huggingface/trl#4612]
I noticed a compatibility issue in GRPOTrainer when using use_liger_loss=True combined with a PEFT (LoRA) model where the lm_head is targeted for training.

In the compute_liger_loss method, the code directly passes the lm_head.weight of the unwrapped model to the LigerFusedLinearGRPOLoss.

```
def compute_liger_loss(self, unwrapped_model, inputs):
    # ...
    last_hidden_state = self._get_last_hidden_state(...)

    # Issue: unwrapped_model.lm_head.weight is the frozen base weight when LoRA is active
    loss, metrics = self.liger_grpo_loss(
        _input=last_hidden_state,
        lin_weight=unwrapped_model.lm_head.weight, 
        selected_token_ids=completion_ids,
        # ...
    )
    # ...
```

If the user configures LoRA with target_modules including "lm_head", the unwrapped_model.lm_head becomes a LoraLayer. 

In this case:
1. unwrapped_model.lm_head.weight refers to the frozen base model weights.
2. The trainable parameters are in the separate LoRA adapters (lora_A, lora_B).
3. Since LigerFusedLinearGRPOLoss computes the operation using only the provided lin_weight, the calculation effectively ignores the LoRA adapters.

The LoRA adapters attached to the lm_head will not contribute to the loss calculation and, consequently, will not receive correct gradient updates. The model will behave as if the lm_head is frozen, even though the user intended to train it.

If lm_head is adapted via PEFT, GRPOTrainer should either:
- Raise a warning or error preventing the usage of use_liger_loss=True when lm_head is in target_modules.
- Or, handle the merging of weights before passing them to the Liger kernel (though this might negate the memory benefits of using Liger).

**Comment by pramodith** [comment]
Interesting scenario! I think raising a warning is the best way to go, we might want to raise an error only if the `target_modules=["lm_head"]` i.e. no other linear layer is being targeted besides the final linear layer.

I'm not sure if merging the weights before passing them to Liger would be numerically correct, because then we'd be optimizing the merged weights and not just the lora adapter since there's no way to distinguish them once the weights have been merged.

I think the ideal solution is for the liger team to add an optional `adaptor_lm_head` or something like that on their end, where we can pass in the adaptor's params and they can handle `lm_head(x) + adaptor_lm_head(x)` internally to ensure gradients and updates are computed correctly.

**Comment by albertvillanova** [comment]
Hey @junho328, sorry for the slow turnaround — catching up on the backlog. 

Confirmed: in the Liger path, `compute_liger_loss` reads `unwrapped_model.lm_head.weight` directly (https://github.com/huggingface/trl/blob/b58947c7e3f6a3ccddc56771f4bfa65c0d72bb41/trl/trainer/grpo_trainer.py#L2382) and Liger runs a raw fused matmul against that `lin_weight`, so when `lm_head` is a LoRA target the frozen base weight is used and the `lora_A`/`lora_B` adapters never enter the graph. I reproduced it on a tiny model: with `lm_head` in `target_modules`, the adapter params get `None` gradients while `lm_head.weight` stays `requires_grad=False`. 

In relation with the proposal by @pramodith to emit a warning, I would argue for always raising an error: 
- The user's intent is the key issue. lm_head appearing in target_modules is an explicit statement: "I want to train this layer." The combination with Liger doesn't produce a partially correct result: it produces a result that silently violates that stated intent. The other adapters training correctly doesn't change the fact that lm_head was supposed to train and didn't.
- Warnings are routinely ignored in ML training pipelines. A user who misses the warning will produce a checkpoint where lm_head was never updated, think it was, and potentially spend days debugging downstream behavior (or worse, never notice).

PR #5977 adds a guard that errors for the `use_liger_loss=True` + LoRA-on-`lm_head` combo, since merging would defeat Liger's memory savings — a clean fix really needs Liger to accept an adapter on the head. I am reviewing it.

Thanks for the precise write-up!

**Comment by namhkoh** [comment]
Great find!
